llvm/lib/SYCLPropagateAspectUsage/PropagateAspectUsage.cpp

```diff
@@ -196,6 +199,76 @@ template <typename T> void PrintMap(const T &m) {
   }
 }
 
+namespace {
+class MissedAspectDiagnosticInfo : public DiagnosticInfo {
+  Twine Msg;
+
+public:
+  MissedAspectDiagnosticInfo(Twine DiagMsg,
+                             DiagnosticSeverity Severity = DS_Warning)
+      : DiagnosticInfo(DK_Linker, Severity), Msg(std::move(DiagMsg)) {}
+
+  void print(DiagnosticPrinter &DP) const override { DP << Msg; }
+};
+} // namespace
+
+void emitWarning(LLVMContext &C, const StringRef Msg) {
+  C.diagnose(MissedAspectDiagnosticInfo(Msg));
+}
+
+template <class Container> std::string Join(const Container &C, char sep) {
+  std::string S;
+  auto it = C.begin();
+  for (size_t i = 0; i != C.size(); ++i, ++it) {
+    if (i > 0)
+      S += ", ";
+
+    S += std::to_string(*it);
+  }
+
+  return S;
+}
+
+template <class Container>
+void CheckDeclaredAspects(LLVMContext &C, const Function *F,
+                          const Container &Aspects) {
+  MDNode *MDN = F->getMetadata("intel_declared_aspects");
+  if (MDN == nullptr)
+    return;
+
+  SmallSet<int, 10> DeclaredAspects;
+  for (unsigned i = 0; i != MDN->getNumOperands(); ++i) {
+    const ConstantAsMetadata *CM =
+        dyn_cast<const ConstantAsMetadata>(MDN->getOperand(i));
+    DeclaredAspects.insert(
+        dyn_cast<const ConstantInt>(CM->getValue())->getSExtValue());
+  }
+
+  SmallVector<int> MissedAspects;
+  for (int aspect : Aspects) {
+    if (DeclaredAspects.count(aspect) == 0) {
+      MissedAspects.push_back(aspect);
+    }
+  }
+
+  if (!MissedAspects.empty()) {
+    std::string Msg;
+    std::string AspectsStr = Join(MissedAspects, ',');
+    // TODO: demangle function name and aspect's IDs?
+    if (MissedAspects.size() > 1) {
+      Msg = formatv("for function \"{0}\" aspects [{1}] are missed in function "
+                    "declaration",
+                    F->getName(), AspectsStr);
+    } else {
+      Msg = formatv(
+          "for function \"{0}\" aspect {1} is missed in function declaration",
+          F->getName(), AspectsStr);
+    }
```

Comment:
I suggest to unify this into something like "The following aspects are missing in function declaration: [list here]"